Support powerful/smart commit messages

sun opened the pull request:
cf. https://drupal.org/node/2270763

Live example:

``` diff
-Issue #2269385 by sun: Fixed DependencyInjection YamlFileLoader is out of date.
+Fixed #2269385 by sun: DependencyInjection YamlFileLoader is out of date.
```

---

~~Problematic / todo:~~

UPDATE: Bogus manual testing result; the issue in question is a task, and "fix*" is only stripped from bug reports.

``` diff
-Issue #2254203 by sun: Fix test performance of Drupal\system\Tests\Module\ModuleApiTest.
+Fixed #2254203 by sun: Fix test performance of Drupal\system\Tests\Module\ModuleApiTest.
```

should be:

``` diff
+Fixed #2254203 by sun: Test performance of Drupal\system\Tests\Module\ModuleApiTest.
```

…can't do anything about poor issue titles like that, but _"Fixed"_ should not be repeated + title _should_ start capitalized.

sun committed:
Changed suggested commit message to support powerful/smart commit messages.

cf. https://drupal.org/node/2270763

sun committed:
Added unit tests.

sun commented:
Sorry, I was mistaken.  This works correctly.

The second issue tested was a task and not a bug report, so the "Fix" prefix in the issue title is not removed, as expected.

That's actually a great use-case for unit tests, because the code logic for generating a commit message does not have any external dependencies.

→ Added fully-fledged unit test coverage.

markhalliwell commented:
I'm ok with merging as is, however I kind feel it should be the verb tense as "fix" implies needed action, not actions already taken (the commit):

`Fixed #N` or `Fixes #N` instead of `Fix #N`, as the commit "fixes" or "fixed" the specified issue.

sun commented:
Note that this is not meant to be merged yet.

The recommended prefix should be discussed on https://drupal.org/node/2270763

markhalliwell commented:
> This works correctly

Ah. Misunderstood what you meant by that. No worries then, I'll comment on that issue.

sun committed:
Changed prefix to be journal-style by default.

sun commented:
Heh.  I just added a commit to change the prefix to be journal-style by default + updated the issue summary here, which revealed:

``` diff
-Issue #2269385 by sun: Fixed DependencyInjection YamlFileLoader is out of date.
+Fixed #2269385 by sun: DependencyInjection YamlFileLoader is out of date.
```

Based on grammar, the past tense doesn't work so well with the "by" line… :collision: ;-)
